torch/linalg/__init__.py

```diff
@@ -488,6 +488,90 @@
+or of each matrix in a batched :attr:`input`.
+The pseudo-inverse is computed using singular value decomposition (see :func:`torch.linalg.svd`) by default.
+If :attr:`hermitian` is ``True``, then :attr:`input` is assumed to be Hermitian (symmetric if real-valued),
+and the computation of the pseudo-inverse is done by obtaining the eigenvalues and eigenvectors
+(see :func:`torch.linalg.eigh`).
+The singular values (or the absolute values of the eigenvalues when :attr:`hermitian` is ``True``) that are below
+the specified :attr:`rcond` threshold are treated as zero and discarded in the computation.
+
+Supports input of ``float``, ``double``, ``cfloat`` and ``cdouble`` datatypes.
+
+.. note:: When given inputs on a CUDA device, this function synchronizes that device with the CPU.
+
+.. note:: If singular value decomposition or eigenvalue decomposition algorithms do not converge
+          then a RuntimeError will be thrown.
+
+Args:
+    input (Tensor): the input matrix of size :math:`(m, n)` or the batch of matrices of size :math:`(*, m, n)`
+                    where `*` is one or more batch dimensions.
+    rcond (float, Tensor, optional): the tolerance value to determine the cutoff for small singular values. Default: 1e-15
+                                     :attr:`rcond` must be broadcastable to singular values of :attr:`input`.
+    hermitian(bool, optional): indicates whether :attr:`input` is Hermitian. Default: ``False``
+
+Examples::
+
+    >>> input = torch.randn(3, 5)
+    >>> input
+    tensor([[ 0.5495,  0.0979, -1.4092, -0.1128,  0.4132],
+            [-1.1143, -0.3662,  0.3042,  1.6374, -0.9294],
+            [-0.3269, -0.5745, -0.0382, -0.5922, -0.6759]])
+    >>> torch.linalg.pinv(input)
+    tensor([[ 0.0600, -0.1933, -0.2090],
+            [-0.0903, -0.0817, -0.4752],
+            [-0.7124, -0.1631, -0.2272],
+            [ 0.1356,  0.3933, -0.5023],
+            [-0.0308, -0.1725, -0.5216]])
+
+    Batched linalg.pinv example
+    >>> a = torch.randn(2, 6, 3)
+    >>> b = torch.linalg.pinv(a)
+    >>> torch.matmul(b, a)
+    tensor([[[ 1.0000e+00,  1.6391e-07, -1.1548e-07],
+            [ 8.3121e-08,  1.0000e+00, -2.7567e-07],
+            [ 3.5390e-08,  1.4901e-08,  1.0000e+00]],
+
+            [[ 1.0000e+00, -8.9407e-08,  2.9802e-08],
+            [-2.2352e-07,  1.0000e+00,  1.1921e-07],
+            [ 0.0000e+00,  8.9407e-08,  1.0000e+00]]])
+
+    Hermitian input example
+    >>> a = torch.randn(3, 3, dtype=torch.complex64)
+    >>> a = a + a.t().conj()  # creates a Hermitian matrix
+    >>> b = torch.linalg.pinv(a, hermitian=True)
+    >>> torch.matmul(b, a)
+    tensor([[ 1.0000e+00+0.0000e+00j, -1.1921e-07-2.3842e-07j,
+            5.9605e-08-2.3842e-07j],
+            [ 5.9605e-08+2.3842e-07j,  1.0000e+00+2.3842e-07j,
+            -4.7684e-07+1.1921e-07j],
+            [-1.1921e-07+0.0000e+00j, -2.3842e-07-2.9802e-07j,
+            1.0000e+00-1.7897e-07j]])
+
+    Non-default rcond example
```

Comment:
This example is tricky. Because we don't see the intermediate steps it's hard to understand the effect that rcond is having. The next example has the same issue.

What are your thoughts? I think we can leave out these examples for now since this PR seems to be almost ready to merge, otherwise, and maybe return to add them later?

Reply:
The documentation says that the pseudo-inverse is calculated using SVD and it says that rcond determines which singular values should be set to zero. The first example demonstrates that using the default and some other value for rcond gives different results as expected. Just using the facts from the docs I think it's understandable that different rconds in general give different results. I think we should keep this example because I think examples should contain the code snippets demonstrating each variable. That's studying math to really understand the effect of rcond (topic of Tikhonov regularization) and definitely not needed here. We're on the level of "something in the input changes, something in the output should change as well".

The purpose of the second rcond example (`Matrix-wise rcond example`) is to show what shapes of rcond are acceptable. I agree that the second example doesn't look transparent and it might be difficult to understand, but that's broadcasting. Smaller arrays are extended to higher-dimensional arrays. We have `rcond must be broadcastable to singular values of input.` in the docs. I don't see how to help the user to understand this bit better.

How about keeping these examples as is (that's working and valid code) and return to it later to improve them?
I don't mind removing it now.

Reply:
@mruberry, so should we remove those rcond examples now and think about it later or keep it?

Reply:
No it's fine, if you like them let's keep them.